bridges/snowbridge/primitives/router/src/outbound/barriers.rs

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+
+use core::{marker::PhantomData, ops::ControlFlow};
+use frame_support::traits::{Contains, ProcessMessageError};
+use xcm::prelude::{ExportMessage, Instruction, Location, NetworkId, Weight};
+use xcm_builder::{CreateMatcher, MatchXcm};
+use xcm_executor::traits::{Properties, ShouldExecute};
+
+pub struct DenyFirstExportMessageFrom<FromOrigin, ToGlobalConsensus>(
+	PhantomData<(FromOrigin, ToGlobalConsensus)>,
+);
+
+impl<FromOrigin, ToGlobalConsensus> ShouldExecute
+	for DenyFirstExportMessageFrom<FromOrigin, ToGlobalConsensus>
+where
+	FromOrigin: Contains<Location>,
+	ToGlobalConsensus: Contains<NetworkId>,
+{
+	fn should_execute<RuntimeCall>(
+		origin: &Location,
+		message: &mut [Instruction<RuntimeCall>],
+		_max_weight: Weight,
+		_properties: &mut Properties,
+	) -> Result<(), ProcessMessageError> {
+		message.matcher().match_next_inst_while(
+			|_| true,
+			|inst| match inst {
+				ExportMessage { network, .. } =>
+					if ToGlobalConsensus::contains(network) && FromOrigin::contains(origin) {
+						return Err(ProcessMessageError::Unsupported)
+					} else {
+						Ok(ControlFlow::Continue(()))
+					},
+				_ => Ok(ControlFlow::Continue(())),
+			},
+		)?;
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use frame_support::{
+		assert_err, parameter_types,
+		traits::{Equals, Everything, EverythingBut},
+	};
+	use xcm::prelude::*;
+	use xcm_builder::{DenyReserveTransferToRelayChain, DenyThenTry, TakeWeightCredit};
+
+	parameter_types! {
+		pub AssetHubLocation: Location = Location::new(1, Parachain(1000));
+		pub EthereumNetwork: NetworkId = Ethereum { chain_id: 1 };
+	}
+
+	#[test]
+	fn deny_export_message_from_source() {
+		let mut xcm: Vec<Instruction<()>> =
+			vec![ExportMessage { network: Polkadot, destination: Here, xcm: Default::default() }];
```

Comment:
```suggestion
			vec![ExportMessage { network: EthereumNetwork::get(), destination: Here, xcm: Default::default() }];
```

?

Test makes more sense like that I think.